openvasd/openvasd.c

```diff
@@ -0,0 +1,2329 @@
+    {
+      g_warning ("%s: Error sending request: %d", __func__, ret);
+      curl_easy_cleanup (curl);
+      response->code = http_code;
+      response->body = g_strdup ("{\"error\": \"Error sending request\"}");
+      return response;
+    }
+
+  curl_easy_getinfo (curl, CURLINFO_RESPONSE_CODE, &http_code);
+  if (header_name)
+    {
+      struct curl_header *hname;
+      curl_easy_header (curl, header_name, 0, CURLH_HEADER, -1, &hname);
+      response->header = g_strdup (hname->value);
+    }
+  curl_easy_cleanup (curl);
+  response->code = http_code;
+
+  return response;
+}
+
+/**
+ * @brief Request HEAD
+ *
+ * @param conn Connector struct with the data necessary for the connection
+ *
+ * @return Response containing the header information
+ */
+openvasd_resp_t
+openvasd_get_version (openvasd_connector_t conn)
+{
+  gchar *err = NULL;
+  CURL *hnd = NULL;
+  openvasd_resp_t response = NULL;
+  struct curl_slist *customheader = NULL;
+
+  response = g_malloc0 (sizeof (struct openvasd_response));
+
+  customheader = init_customheader (conn->apikey, FALSE);
+  if ((hnd = handler (conn, HEAD, "/", NULL, customheader, &err)) == NULL)
+    {
+      curl_slist_free_all (customheader);
+      response->code = RESP_CODE_ERR;
+      response->body = err;
+      openvasd_reset_vt_stream (conn);
+      return response;
+    }
+
+  openvasd_send_request (hnd, NULL, response);
+  curl_slist_free_all (customheader);
+  if (response->code != RESP_CODE_ERR)
+    response->body = g_strdup (openvasd_vt_stream_str (conn));
+
+  openvasd_reset_vt_stream (conn);
+  return response;
+}
+
+/**
+ * @brief Initialized an curl multiperform handler which allows fetch feed
+ * metadata chunk by chunk.
+ *
+ * @param conn Connector struct with the data necessary for the connection
+ * @param mhnd The curl multiperform handler. It the caller doesn't provide
+ * it initialized, it will be initialized. The caller has to free it with
+ * openvasd_curlm_handler_close().
+ * @param resp The stringstream struct for the write callback function.
+ *
+ * @return The response.
+ */
+openvasd_resp_t
+openvasd_get_vt_stream_init (openvasd_connector_t conn)
+{
+  GString *path;
+  openvasd_resp_t response = NULL;
+  gchar *err = NULL;
+  CURL *hnd = NULL;
+  CURLM *h = NULL;
+  struct curl_slist *customheader = NULL;
+
+  response = g_malloc0 (sizeof (struct openvasd_response));
+
+  path = g_string_new ("/vts?information=1");
+  customheader = init_customheader (conn->apikey, FALSE);
+  if ((hnd = handler (conn, GET, path->str, NULL, customheader, &err)) == NULL)
+    {
+      curl_slist_free_all (customheader);
+      g_string_free (path, TRUE);
+      response->code = RESP_CODE_ERR;
+      response->body = err;
+      return response;
+    }
+  g_string_free (path, TRUE);
+
+  h = curl_multi_init ();
+  curl_multi_add_handle (h, hnd);
+
+  if (conn->stream_resp->curl_hnd->h == NULL)
+    conn->stream_resp->curl_hnd->h = openvasd_curlm_handler_new ();
```

Comment:
This should be working with `curl_hnd` instead of `curl_hnd->h`.  The `->h` is assigned below.

Maybe just remove these two lines? The curl_hnd is always created by `openvas_vt_stream_new`, which is static.

Reply:
What about this @jjnicola?

See how `conn->stream_resp->curl_hnd->h` is assigned 2 lines later.